design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,108 @@
+# Overview
+This proposal introduces a mechanism to make the firmware UUID of a Virtual Machine in KubeVirt universally unique.
+It ensures the UUID remains consistent across VM restarts, preserving stability and reliability.
+
+## Motivation
+* Duplicate UUIDs cause burden with third-party automation frameworks such as Gitops tools, making it difficult to maintain unique VM UUID. 
+  Such tools have to manually assign unique UUIDs to VMs, thus also manage them externally. (issue)
+* The non-persistent nature of UUIDs may trigger redundant cloud-init setup upon VM restore. (issue)
+* The non-persistent nature of UUIDs may trigger redundant license activation events during VM lifetime i.e. Windows key activation re-occur when power off and power on the VM.
+
+## Goals
+* Improve the uniqueness of the VMI firmware UUID to become independent of the VM name and namespace.
+* Maintain the UUID persistence over the VM lifecycle.
+* Maintain backward compatibility
+
+## Non Goals
+* Maintain UUID persistence with VM backup/restore.
+* Change UUID of currently existing VMs
+
+## Definition of Users
+VM owners: who require consistent firmware UUIDs for their applications.
+cluster-admin: to ensure VMs have universally unique firmware IDs
+
+## User Stories
+### Supported cases:
+* Creating and Starting a New VM
+* starting existing VM that was not running
+
+
+## Repos
+Kubevirt/kubevirt
+
+
+## Proposed Solution
+
+### Description
+The firmware UUID persistence mechanism will operate as follows:
+
+1. **New VMs**:
+    - If the firmware UUID is not explicitly defined in `vm.spec.template.spec.firmware.uuid`, the mutator webhook will automatically set the firmware UUID to the value of `vm.metadata.uid`.
+
+2. **Existing VMs**:
+    - For VMs created before the upgrade, the VM controller will patch the `vm.spec.template.spec.firmware.uuid` field,
+   with a value calculated using the legacy logic (based on the VM name).
+   This ensures backward compatibility.
+
+3. **Mitigation**:
+    - Backups created before the implementation of this mechanism will result in VMs receiving a new UUID upon restore, as the firmware UUID field will be absent in older backups.
+    - Users must be aware of this limitation and plan accordingly.
```

Comment:
can we hint here what they should be doing?
It would be nice if we can have a script that adds the buggy firmware UUID on existing snapshots.